### PR TITLE
feat(#361):report real-time build status while the pipelines are running

### LINKED ISSUE #361: Report build status while the pipelines are running
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
Now the pipelines only report status when they finish.

**What you expected to happen**:
Report the status in real-time while the pipelines are running.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:
/assign @zhujian7

### PR DESCRIPTION
**Pull request by zhujian7**
pipelines only report status when they finish

**What this PR does / why we need it**:

report real-time build status while the pipelines are running

**Which issue(s) this PR fixes** *(optional, close the issue(s) when PR gets merged)*:

Fixes #361 

**Special notes for your reviewer**:

/cc @supereagle 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/server/manager/event.go [hunk 1]
```diff
@@ -55,12 +56,13 @@ func (m *eventManager) GetEvent(id string) (*api.Event, error) {
...
+func (m *eventManager) SetEvent(e *api.Event) (*api.Event, error) {
+	err := m.ds.UpdatePipelineRecord(e.PipelineRecord)
 	if err != nil {
-		log.Errorf("Fail to set the event %s", event.ID)
+		log.Errorf("Fail to set the pipeline record %s", e.PipelineRecord.ID)
 		return nil, err
 	}
 
-	return event, nil
+	event.UpdateEvent(e)
```
Comment (supereagle) [inline pkg/server/manager/event.go]: No need to check the result?

File: pkg/worker/stage/stage.go [hunk 2]
```diff
@@ -110,12 +140,41 @@ func (sm *stageManager) ExecCodeCheckout(token string, stage *api.CodeCheckoutSt
...
+		if err != nil {
+			event.PipelineRecord.StageStatus.Package.Status = api.Failed
+			event.PipelineRecord.StageStatus.Package.EndTime = time.Now()
+			event.PipelineRecord.Status = api.Failed
+			event.PipelineRecord.ErrorMessage = fmt.Sprintf("package fail : s%", err.Error())
+			go sm.cycloneClient.SetEvent(event)
+		} else {
+			event.PipelineRecord.StageStatus.Package.Status = api.Success
+			event.PipelineRecord.StageStatus.Package.EndTime = time.Now()
+			go sm.cycloneClient.SetEvent(event)
```
Comment (supereagle) [inline pkg/worker/stage/stage.go]: Move out of if-else.

File: pkg/worker/stage/stage.go [hunk 3]
```diff
@@ -191,6 +250,29 @@ func (sm *stageManager) ExecPackage(builderImage *api.BuilderImage, unitTestStag
 }
 
 func (sm *stageManager) ExecImageBuild(stage *api.ImageBuildStage) ([]string, error) {
+	var err error
+	event.PipelineRecord.StageStatus.ImageBuild = &api.GeneralStageStatus{
+		//		GeneralStageStatus: api.GeneralStageStatus{
+		Status:    api.Running,
+		StartTime: time.Now(),
+		//		},
```
Comment (supereagle) [inline pkg/worker/stage/stage.go]: Remove.

File: pkg/worker/stage/stage.go [hunk 4]
```diff
@@ -78,7 +81,32 @@ func (sm *stageManager) SetRecordInfo(project, pipeline, recordID string) {
 	sm.recordID = recordID
 }
 
-func (sm *stageManager) ExecCodeCheckout(token string, stage *api.CodeCheckoutStage) error {
+func (sm *stageManager) SetEvent(e *api.Event) {
```
Comment (supereagle) [inline pkg/worker/stage/stage.go]: Why need this function? We can directly set the event.

File: pkg/worker/stage/stage.go [hunk 5]
```diff
@@ -110,12 +138,37 @@ func (sm *stageManager) ExecCodeCheckout(token string, stage *api.CodeCheckoutSt
 	if err != nil {
 		return err
 	}
+	formatPipelineRecordName(commitID)
+
+	repoName, errn := scm.GetRepoName(codeSource)
+	if errn != nil {
+		log.Warningf("get repo name fail %s", errn.Error())
+	}
+	commitLog := scmProvider.GetTagCommitLog(cloneDir, "master")
```
Comment (supereagle) [inline pkg/worker/stage/stage.go]: We should support other branches and tags.

File: pkg/worker/stage/stage.go [hunk 6]
```diff
@@ -78,7 +81,32 @@ func (sm *stageManager) SetRecordInfo(project, pipeline, recordID string) {
...
+
+// ExecCodeCheckout Checkout code and report real-time status to cycloue server.
+func (sm *stageManager) ExecCodeCheckout(token string, stage *api.CodeCheckoutStage) (err error) {
+	event.PipelineRecord.StageStatus.CodeCheckout = &api.CodeCheckoutStageStatus{
+		GeneralStageStatus: api.GeneralStageStatus{
+			Status:    api.Running,
+			StartTime: time.Now(),
+		},
+	}
+	go sm.cycloneClient.SendEvent(event)
```
Comment (supereagle) [inline pkg/worker/stage/stage.go]: Not need a goroutine to send event.

File: pkg/worker/stage/stage.go (hunk 4 above)
Comment (zhujian7) [reply pkg/worker/stage/stage.go]: Authority control，other package can not get the event(variable e).

File: pkg/worker/stage/stage.go (hunk 5 above)
Comment (zhujian7) [reply pkg/worker/stage/stage.go]: done

File: pkg/worker/stage/stage.go (hunk 6 above)
Comment (zhujian7) [reply pkg/worker/stage/stage.go]: done

**Pull request merged**